# Dynamic Programming to Determine the Count of Multiple Solutions with 1, 2, and 3 Steps

## When trying to figure out how to get to cover distances there are multiple approaches that can be taken. For instance to get to a distance of 3, it can be covered in 3 one steps, 1 one step and 1 two step, 1 3 step, and 1 two step and 1 one step. This makes for a total of 4 methods that it can take to cover a distance of 3. Using similar methods to cover a distance of 4, there are a total of 7 methods to get there when limited at most to 3 steps. The number of methods it can be used to cover a certain distance gets more outrageous as time goes on, to the point where it would be nearly impossible to cover by hand. 

## However, through the use of dynamic programming figuring out how many methods that can be used to cover outrageous distances with only 3 step, 2 step or 1 step intervals is a fairly simple process, below is an instance of using dynamic programming as a solution. 

In [14]:
def printCountDP(dist):
    count = [0] * (dist + 1)
    count[0] = 1
    if dist >= 1 :
        count[1] = 1
    if dist >= 2 :
        count[2] = 2
    for i in range(3, dist + 1):
        count[i] = (count[i-1] +
                   count[i-2] + count[i-3])
    print(count)     
    return count[dist];

printCountDP(10)

[1, 1, 2, 4, 7, 13, 24, 44, 81, 149, 274]


274

## This function is a dynamic program since it solves each distance and continues to build off of the previous distance. Here is how the program works.

### 1. The program creates an array of zeroes. This array is based off of the distance needed to cover plus 1 (this is to prevent no array being created if the distance is 0).

### 2. The program sets the first array position to 1. If the distance is equal or greater than 1 then it sets the second position to 1. If the distance is greater than or equal to 2 it sets the 3rd position equal to 2.

### 3. From there a for loop is created out of a range from 3 to the distance plus 1 (to make sure that final distance number is covered), each position is therefore set equal to the count of the previous step plus the count of the step before the previous plus the step before the step before the prevous set.

### 4. What is return is the value in the position of where the distance is.

## The function is a dynamic program since it breaks down each distance and uses that distance as a substep to getting to the final solution. For instance, when performing the function with the distance 10 there are 274 methods received. This is based off of adding the subset values 44, 81, and 149 (and even these are based off of solutions from previous steps). This dynamic programming solution basically allows for further build up over time.

## This dynamic programming has a Big O notation of O(n) meaning that the program will increase at a fairly linear rate. As the size of the distance increases the time it will take to complete this program will increase as well. It is important for every data engineer to utilize dynamic programming since it is capable of taking breaking down problems into subsets of solution. This, at times, will allow it to run at a faster rate. It is also important to find a dynamic programming solution that is optimal for Big O notation. There is a solution to this form that would have had a Big O notation of O(3^n) which is a far slower rate in comparison to the above problem implemented.